src/config.rs

```diff
@@ -717,23 +717,196 @@ pub mod log_level_serde {
     }
 }
 
-#[derive(Clone, Serialize, Deserialize, PartialEq, Debug)]
-#[serde(default)]
-#[serde(rename_all = "kebab-case")]
-pub struct ReadPoolConfig {
-    pub storage: ReadPoolInstanceConfig,
-    pub coprocessor: ReadPoolInstanceConfig,
+macro_rules! readpool_config {
+    ($struct_name:ident, $test_mod_name:ident, $display_name:expr) => {
+        #[derive(Clone, Serialize, Deserialize, PartialEq, Debug)]
+        #[serde(default)]
+        #[serde(rename_all = "kebab-case")]
+        pub struct $struct_name {
+            pub high_concurrency: usize,
+            pub normal_concurrency: usize,
+            pub low_concurrency: usize,
+            pub max_tasks_high: usize,
+            pub max_tasks_normal: usize,
+            pub max_tasks_low: usize,
+            pub stack_size: ReadableSize,
+        }
+
+        impl $struct_name {
+            pub fn build_config(&self) -> readpool::Config {
+                readpool::Config {
+                    high_concurrency: self.high_concurrency,
+                    normal_concurrency: self.normal_concurrency,
+                    low_concurrency: self.low_concurrency,
+                    max_tasks_high: self.max_tasks_high,
+                    max_tasks_normal: self.max_tasks_normal,
+                    max_tasks_low: self.max_tasks_low,
+                    stack_size: self.stack_size,
+                }
+            }
+
+            pub fn validate(&self) -> Result<(), Box<Error>> {
```

Comment:
because at the time when calling `validate`, we have not converted it to `readpool::Config` yet.

Reply:
Note that, during the lifetime of config, it is always `StorageReadPoolConfig` and `CoprocessorReadPoolConfig`. Only when we creating read pools, we derive the config using `build_config` to get an acceptable structure. This is the main reason why we need to validate in `XxxReadPoolConfig` instead of `readpool::Config`.